### ISSUE TITLE
XS layout/Makefile.PL configuration leads to repeated builds

### ISSUE THREAD

**Issue by kentfredric** [houseabsolute/File-LibMagic#28]
This problem is easier to demonstrate if you check out 1.03 and see what it doesn't do:

```bash
# /usr/bin/perl Makefile.PL 
Checking for magic.h... yes
Checking for magic_open in -lmagic... yes
Checking if your kit is complete...
Looks good
Generating a Unix-style Makefile
Writing Makefile for File::LibMagic
Writing MYMETA.yml and MYMETA.json
# make OPTIMIZE="-O3 -frecord-gcc-switches"
cp lib/File/LibMagic.pm blib/lib/File/LibMagic.pm
Running Mkbootstrap for LibMagic ()
chmod 644 "LibMagic.bs"
"/usr/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- LibMagic.bs blib/arch/auto/File/LibMagic/LibMagic.bs 644
"/usr/bin/perl" "/usr/lib64/perl5/5.32/ExtUtils/xsubpp"  -typemap '/usr/lib64/perl5/5.32/ExtUtils/typemap'  LibMagic.xs > LibMagic.xsc
mv LibMagic.xsc LibMagic.c
x86_64-pc-linux-gnu-gcc -c  -I. -fwrapv -fno-strict-aliasing -pipe -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -O3 -frecord-gcc-switches   -DVERSION=\"1.03\" -DXS_VERSION=\"1.03\" -fPIC "-I/usr/lib64/perl5/5.32/x86_64-linux/CORE"   LibMagic.c
rm -f blib/arch/auto/File/LibMagic/LibMagic.so
LD_RUN_PATH="/usr/lib64" x86_64-pc-linux-gnu-gcc  -shared -O2 -pipe -mtune=native -march=native -fstack-protector-strong -fno-stack-protector -Wl,-O1  -Wl,--as-needed -fstack-protector-strong -fno-stack-protector  LibMagic.o  -o blib/arch/auto/File/LibMagic/LibMagic.so  \
   -lmagic   \
  
chmod 755 blib/arch/auto/File/LibMagic/LibMagic.so
Manifying 1 pod document
# make
"/usr/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- LibMagic.bs blib/arch/auto/File/LibMagic/LibMagic.bs 644
Manifying 1 pod document
```

The key observation here is subsequent calls to `make` don't rebuild the `.so` components, and thus, the `.so` components retain the flags that were passed in the initial `make` call.

Now, compare with `v1.22`
```bash
# /usr/bin/perl Makefile.PL 
Checking for magic.h... Checking for cc... x86_64-pc-linux-gnu-gcc
yes
Checking for magic_open in -lmagic... yes
Checking for magic_version in -lmagic... yes
Checking for magic_setparam in -lmagic... yes
Checking for magic_getparam in -lmagic... yes
Checking if your kit is complete...
Looks good
Generating a Unix-style Makefile
Writing Makefile for File::LibMagic
Writing MYMETA.yml and MYMETA.json
# make OPTIMIZE="-O3 -frecord-gcc-switches"
cp lib/File/LibMagic.pm blib/lib/File/LibMagic.pm
cp lib/File/LibMagic/Constants.pm blib/lib/File/LibMagic/Constants.pm
Running Mkbootstrap for LibMagic ()
chmod 644 "LibMagic.bs"
"/usr/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- LibMagic.bs blib/arch/auto/File/LibMagic/LibMagic.bs 644
"/usr/bin/perl" "/usr/lib64/perl5/5.32/ExtUtils/xsubpp"  -typemap '/usr/lib64/perl5/5.32/ExtUtils/typemap'  lib/File/LibMagic.xs > lib/File/LibMagic.xsc
mv lib/File/LibMagic.xsc lib/File/LibMagic.c
x86_64-pc-linux-gnu-gcc -c  -I. -Ic -fwrapv -fno-strict-aliasing -pipe -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -O3 -frecord-gcc-switches   -DVERSION=\"1.22\" -DXS_VERSION=\"1.22\" -fPIC "-I/usr/lib64/perl5/5.32/x86_64-linux/CORE"  -DHAVE_MAGIC_VERSION -DHAVE_MAGIC_SETPARAM -DHAVE_MAGIC_GETPARAM lib/File/LibMagic.c
rm -f blib/arch/auto/File/LibMagic/LibMagic.so
LD_RUN_PATH="/usr/lib64" x86_64-pc-linux-gnu-gcc  -shared -O2 -pipe -mtune=native -march=native -fstack-protector-strong -fno-stack-protector -Wl,-O1 -Wl,--as-needed -fstack-protector-strong -fno-stack-protector  LibMagic.o  -o blib/arch/auto/File/LibMagic/LibMagic.so  \
   -lmagic   \
  
chmod 755 blib/arch/auto/File/LibMagic/LibMagic.so
Manifying 2 pod documents
# make
"/usr/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- LibMagic.bs blib/arch/auto/File/LibMagic/LibMagic.bs 644
x86_64-pc-linux-gnu-gcc -c  -I. -Ic -fwrapv -fno-strict-aliasing -pipe -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -O2 -pipe -mtune=native -march=native -fstack-protector-strong -fno-stack-protector   -DVERSION=\"1.22\" -DXS_VERSION=\"1.22\" -fPIC "-I/usr/lib64/perl5/5.32/x86_64-linux/CORE"  -DHAVE_MAGIC_VERSION -DHAVE_MAGIC_SETPARAM -DHAVE_MAGIC_GETPARAM lib/File/LibMagic.c
rm -f blib/arch/auto/File/LibMagic/LibMagic.so
LD_RUN_PATH="/usr/lib64" x86_64-pc-linux-gnu-gcc  -shared -O2 -pipe -mtune=native -march=native -fstack-protector-strong -fno-stack-protector -Wl,-O1 -Wl,--as-needed -fstack-protector-strong -fno-stack-protector  LibMagic.o  -o blib/arch/auto/File/LibMagic/LibMagic.so  \
   -lmagic   \
  
chmod 755 blib/arch/auto/File/LibMagic/LibMagic.so
Manifying 2 pod documents
# make
"/usr/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- LibMagic.bs blib/arch/auto/File/LibMagic/LibMagic.bs 644
x86_64-pc-linux-gnu-gcc -c  -I. -Ic -fwrapv -fno-strict-aliasing -pipe -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -O2 -pipe -mtune=native -march=native -fstack-protector-strong -fno-stack-protector   -DVERSION=\"1.22\" -DXS_VERSION=\"1.22\" -fPIC "-I/usr/lib64/perl5/5.32/x86_64-linux/CORE"  -DHAVE_MAGIC_VERSION -DHAVE_MAGIC_SETPARAM -DHAVE_MAGIC_GETPARAM lib/File/LibMagic.c
rm -f blib/arch/auto/File/LibMagic/LibMagic.so
LD_RUN_PATH="/usr/lib64" x86_64-pc-linux-gnu-gcc  -shared -O2 -pipe -mtune=native -march=native -fstack-protector-strong -fno-stack-protector -Wl,-O1 -Wl,--as-needed -fstack-protector-strong -fno-stack-protector  LibMagic.o  -o blib/arch/auto/File/LibMagic/LibMagic.so  \
   -lmagic   \
  
chmod 755 blib/arch/auto/File/LibMagic/LibMagic.so
Manifying 2 pod documents
```

In its current incarnation, it seems whatever is done in Makefile.PL to faciliate this "have XS inside lib/" strategy, results in `make` cache invalidation fail entirely, and so, the target is recompiled *EVERY TIME* `make` is called.

And subsequently, it breaks when tooling applies to it that didn't expect this behaviour, because they expected the assets to not be unhelpfully recompiled without the required settings.

And its *generally* frowned upon to be recompiling your libraries during `make test` or `make install`.

A certain somebody says:
```irclog
<___> The value for OBJECT is wrong
...
<___> It recompiles because lib/File/LibMagic\$(OBJ_EXT) never exists
...
<___> The whole setup is broken and wrong
<___> I don't recall any way to make .xs in lib/ work without using mohawks new stuff, and even then I recall running into some issues (but forgot details)
```

**Comment by autarch** [comment]
I'm not sure who the certain somebody is but it looks like a trivial fix. See 83553999b92483db9e1a471c0727f269f7664573. Note that for you to actually test this yourself you'd need to run `dzil build` to get an updated `Makefile.PL`. But I tested before and after myself and this did the trick.

**Comment by autarch** [comment]
Doh, spoke too soon. This failed CI and it turns out it only works if the directory is dirty from a previous a build. But somehow Moose makes having xs code in a dir that's not root work, and it doesn't recompile each time.

**Comment by autarch** [comment]
This should be fixed in 1.28.
